File: tests/fixtures/python/bad.py
def ProcessData(d, t, x, flg, optA, optB, optC, optD):
    r = []
    if t == "csv":
        if flg:
            for i in range(len(d)):
                if d[i] is not None:
                    if isinstance(d[i], str):
                        if len(d[i]) > 0:
                            if x:
                                v = d[i].strip().lower()
                                if v not in r:
                                    r.append(v)
                            else:
                                v = d[i].strip()
                                if v not in r:
                                    r.append(v)
        else:
            for i in range(len(d)):
                if d[i] is not None:
                    if isinstance(d[i], str):
                        if len(d[i]) > 0:
                            v = d[i].strip()
                            if v not in r:
                                r.append(v)
    elif t == "json":
        if flg:
            for i in range(len(d)):
                if d[i] is not None:
                    if isinstance(d[i], dict):
                        for k in d[i]:
                            if d[i][k] is not None:
                                if isinstance(d[i][k], str):
                                    v = d[i][k].strip().lower()
                                    if v not in r:
                                        r.append(v)
        else:
            for i in range(len(d)):
                if d[i] is not None:
                    if isinstance(d[i], dict):
                        for k in d[i]:
                            if d[i][k] is not None:
                                if isinstance(d[i][k], str):
                                    v = d[i][k].strip()
                                    if v not in r:
                                        r.append(v)
    elif t == "xml":
        if flg:
            for i in range(len(d)):
                if d[i] is not None:
                    if hasattr(d[i], 'text'):
                        if d[i].text is not None:
                            if len(d[i].text) > 0:
                                v = d[i].text.strip().lower()
                                if v not in r:
                                    r.append(v)
        else:
            for i in range(len(d)):
                if d[i] is not None:
                    if hasattr(d[i], 'text'):
                        if d[i].text is not None:
                            if len(d[i].text) > 0:
                                v = d[i].text.strip()
                                if v not in r:
                                    r.append(v)

    if optA:
        r = sorted(r)
    if optB:
        r = r[:optB]
    if optC:
        r = [x for x in r if len(x) > optC]
    if optD:
        r = [x.upper() for x in r]

    return r


def calc(a,b,c,d,e,f,g,h):
    z = a + b
    z = z * c
    z = z - d
    if e:
        z = z / e
    if f:
        if g:
            if h:
                z = z ** h
                if z > 1000:
                    if z > 10000:
                        z = 10000
                    else:
                        z = z
    return z


class mgr:
    def __init__(s, db, lg, cf, st):
        s.db = db
        s.lg = lg
        s.cf = cf
        s.st = st

    def p(s, d):
        if s.cf.get('v'):
            s.lg.info(d)
        s.db.save(d)
        s.st.update(d)
        return True

    def g(s, id):
        r = s.db.find(id)
        if r:
            s.lg.info(f"found {id}")
        return r


def bad_error_handling(data):
    try:
        result = data["key"]
    except:
        pass

    try:
        value = int(data)
    except Exception:
        value = 0

    try:
        x = 42 / data
    except:
        ...

    return 0


def magic_everywhere(items):
    if len(items) > 37:
        items = items[:37]
    total = 0
    for item in items:
        total += item * 3.14
        if total > 9999:
            total = total / 7
    return total + 256


def wrapped_body(data):
    if data is not None:
        cleaned = data.strip()
        validated = len(cleaned) > 0
        if validated:
            processed = cleaned.lower()
            return processed
        return ""

File: tests/fixtures/php/bad.php
<?php

function ProcessData($d, $t, $x, $flg, $optA, $optB, $optC, $optD)
{
    $r = [];
    if ($t == "csv") {
        if ($flg) {
            for ($i = 0; $i < count($d); $i++) {
                if ($d[$i] !== null) {
                    if (is_string($d[$i])) {
                        if (strlen($d[$i]) > 0) {
                            if ($x) {
                                $v = strtolower(trim($d[$i]));
                                if (!in_array($v, $r)) {
                                    $r[] = $v;
                                }
                            } else {
                                $v = trim($d[$i]);
                                if (!in_array($v, $r)) {
                                    $r[] = $v;
                                }
                            }
                        }
                    }
                }
            }
        } else {
            for ($i = 0; $i < count($d); $i++) {
                if ($d[$i] !== null) {
                    if (is_string($d[$i])) {
                        if (strlen($d[$i]) > 0) {
                            $v = trim($d[$i]);
                            if (!in_array($v, $r)) {
                                $r[] = $v;
                            }
                        }
                    }
                }
            }
        }
    } elseif ($t == "json") {
        if ($flg) {
            for ($i = 0; $i < count($d); $i++) {
                if ($d[$i] !== null) {
                    if (is_array($d[$i])) {
                        foreach ($d[$i] as $k => $val) {
                            if ($val !== null) {
                                if (is_string($val)) {
                                    $v = strtolower(trim($val));
                                    if (!in_array($v, $r)) {
                                        $r[] = $v;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    if ($optA) {
        sort($r);
    }
    if ($optB) {
        $r = array_slice($r, 0, $optB);
    }
    if ($optC) {
        $r = array_filter($r, function($x) use ($optC) {
            return strlen($x) > $optC;
        });
    }
    if ($optD) {
        $r = array_map('strtoupper', $r);
    }

    return $r;
}


function calc($a, $b, $c, $d, $e, $f, $g, $h)
{
    $z = $a + $b;
    $z = $z * $c;
    $z = $z - $d;
    if ($e) {
        $z = $z / $e;
    }
    if ($f) {
        if ($g) {
            if ($h) {
                $z = pow($z, $h);
                if ($z > 1000) {
                    if ($z > 10000) {
                        $z = 10000;
                    } else {
                        $z = $z;
                    }
                }
            }
        }
    }
    return $z;
}


class mgr
{
    private $db;
    private $lg;
    private $cf;
    private $st;

    public function __construct($db, $lg, $cf, $st)
    {
        $this->db = $db;
        $this->lg = $lg;
        $this->cf = $cf;
        $this->st = $st;
    }

    public function p($d)
    {
        if ($this->cf['v']) {
            $this->lg->info($d);
        }
        $this->db->save($d);
        $this->st->update($d);
        return true;
    }

    public function g($id)
    {
        $r = $this->db->find($id);
        if ($r) {
            $this->lg->info("found {$id}");
        }
        return $r;
    }
}


function bad_error_handling($data)
{
    try {
        $result = $data["key"];
    } catch (\Exception $e) {
        // empty catch - swallows exception
    }

    try {
        $value = (int)$data;
    } catch (\Exception $e) {
        $value = 0;
    }

    try {
        $x = 42 / $data;
    } catch (\Throwable $e) {
    }

    return 0;
}


function legacy_code($data)
{
    $conn = mysql_connect("localhost", "root", "");
    mysql_select_db("mydb", $conn);
    $result = mysql_query("SELECT * FROM users");
    $row = mysql_fetch_assoc($result);
    mysql_close($conn);

    $clean = ereg_replace("[^a-z]", "", $data);
    $parts = split(",", $data);

    $fn = create_function('$a', 'return $a * 2;');

    $encoded = utf8_encode($data);
    $date = strftime("%Y-%m-%d");

    return $row;
}


function magic_everywhere($items)
{
    if (count($items) > 37) {
        $items = array_slice($items, 0, 37);
    }
    $total = 0;
    foreach ($items as $item) {
        $total += $item * 3.14;
        if ($total > 9999) {
            $total = $total / 7;
        }
    }
    return $total + 256;
}


function wrappedBody($data)
{
    if ($data !== null) {
        $cleaned = trim($data);
        $validated = strlen($cleaned) > 0;
        if ($validated) {
            $processed = strtolower($cleaned);
            return $processed;
        }
        return "";
    }
}
